File: src/info/scanner.py
import subprocess
import json

# Define the Node.js script's path
node_script_path = 'src\\info\\chart\\scanner.js'


def get_scanner_data(address, chain_id, project, liquidity, holder, score):
    # Execute the Node.js script using subprocess
    process = subprocess.run(['node', node_script_path, address, str(chain_id), project, liquidity, holder, score], capture_output=True, text=True)

    # Check if the subprocess ran successfully
    if process.returncode == 0:
        # Get the stdout from the Node.js script and parse the JSON
        output = process.stdout
        data = json.loads(output)  # Parse the JSON output from Node.js
        # print("Data received from Node.js:", data)
        return data
    else:
        # print("Node.js script failed with return code:", process.returncode)
        return False

def get_scanner_general_result(address, chain_id, liquidity):
    if liquidity:
        data = get_scanner_data(address=address, chain_id=chain_id, project="T", liquidity="T", holder="T", score= "T")
    else:
        data = get_scanner_data(address=address, chain_id=chain_id, project="T", liquidity="F", holder="F", score= "T")
    if data == False or data["project"] == None:
        return None
    else:
        pass
    project_name = data["project"]["name"]
    airealm_score = data["score"]["score"]
    contractname = data["project"]["contractName"]
    issues = []
    for i in data["project"]["coreIssues"]:
        if i["issues"] != []:
            issues.append(i)
        else:
            pass
    if data["holderAnalysis"]:
        for i in data["holderAnalysis"]["issues"]:
            if i["issues"] != []:
                issues.append(i)
            else:
                pass
        
        owner = data["holderAnalysis"]["owner"]
        ownerbalance = data["holderAnalysis"]["ownerBalance"]
        ownerBalancePercentage = data["holderAnalysis"]["ownerBalancePercentage"]

        creator = data["holderAnalysis"]["creator"]
        creatorbalance =  data["holderAnalysis"]["creatorBalance"]
        creatorBalancePercentage = data["holderAnalysis"]["creatorBalancePercentage"]

        supply = data["holderAnalysis"]["tokenTotalSupply"]
    
    else:
        owner = None
        ownerbalance = None
        ownerBalancePercentage = None

        creator = None
        creatorbalance =  None
        creatorBalancePercentage = None

        supply = None

    if data["liquidityAnalysis"]:
        for i in data["liquidityAnalysis"]["issues"]:
            if i["issues"] != []:
                issues.append(i)
            else:
                pass

        liquidity = float(data["liquidityAnalysis"]["totalLiquidity"])
        totalLockedPercent = data["liquidityAnalysis"]["totalLockedPercent"]
        totalUnlockedPercent = data["liquidityAnalysis"]["totalUnlockedPercent"]
    else:
        liquidity = None
        totalLockedPercent = None
        totalUnlockedPercent = None

    return_value = {
        "project_name": project_name,
        "airealm_score": airealm_score, 
        "contractname": contractname,
        "issues": issues, 
        "owner": owner, 
        "ownerbalance": ownerbalance,
        "ownerBalancePercentage": ownerBalancePercentage,
        "creator": creator, 
        "creatorbalance": creatorbalance,
        "creatorBalancePercentage": creatorBalancePercentage,
        "supply": supply, 
        "totalLockedPercent": totalLockedPercent,
        "totalUnlockedPercent": totalUnlockedPercent,
        "liquidity": liquidity
    }
    return return_value

def get_scanner_issues_result(address, chain_id, liquidity):
    if liquidity:
        data = get_scanner_data(address=address, chain_id=chain_id, project="T", liquidity="T", holder="T", score="F")
    else:
        data = get_scanner_data(address=address, chain_id=chain_id, project="T", liquidity="F", holder="F", score="F")
    
    if data == False or data["project"] == None:
        return None
    else:
        pass
    issues = []
    for i in data["project"]["coreIssues"]:
        if i["issues"] != []:
            issues.append(i)
        else:
            pass
    if data["holderAnalysis"]:
        for i in data["holderAnalysis"]["issues"]:
            if i["issues"] != []:
                issues.append(i)
            else:
                pass
    if data["liquidityAnalysis"]:
        for i in data["liquidityAnalysis"]["issues"]:
            if i["issues"] != []:
                issues.append(i)
            else:
                pass

    return_value = {
        "issues": issues
    }
    return return_value

def get_scanner_liquidity_result(address, chain_id):
    data = get_scanner_data(address=address, chain_id=chain_id, project="F", liquidity="T", holder="F", score="F")
    return data["liquidityAnalysis"]

def get_scanner_holders_result(address, chain_id):
    data = get_scanner_data(address=address, chain_id=chain_id, project="F", liquidity="F", holder="T", score="F")
    return data["holderAnalysis"]


# get_scanner_data("0x2170ed0880ac9a755fd29b2688956bd959f933f8", '56')

File: src/info/chart/scanner.js
const puppeteer = require("puppeteer");
const fs = require("fs");

const address = process.argv[2]
const chain_id = Number(process.argv[3])
const project_request = process.argv[4]
const liquidity_request = process.argv[5]
const holder_request = process.argv[6]
const score_request = process.argv[7]

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function fetchGraphQLData(query, variables) {
    const response = await fetch('https://api-scanner.defiyield.app/', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        // Include any other headers you need to send
      },
      body: JSON.stringify({
        query,
        variables
      })
    });
  
    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`);
    }
  
    const data = await response.json();
    return data;
  }
  
const project_query = `
  query($address: String!, $network: Int!) {
    project(address: $address, networkId: $network) {
      inProgress
      isProxyImplementation
      proxyContractAddress
      firstTxBlock
      estimatedAnalyzingTime
      address
      network
      rektLink
      name
      contractName
      sourceCodeLink
      whitelisted
      firstTxFrom
      firstTxDate
      logo
      projectName
      projectFullName
      initialFunder
      protocol
      pairInfo {
        factory
        tokens {
          address
          name
          symbol
          score
          logo
        }
      }
      coreIssues {
        scwTitle
        scwDescription
        scwId
        issues {
          id
          impact
          description
          data
          snippet
          start
          end
          additionalData {
            title
            description
          }
          governanceInfo {
            owners {
              timelockDelay
              type
              owner
            }
            worstOwner {
              timelockDelay
              type
              owner
            }
          }
          severityChanges {
            from
            to
            reason
          }
        }
      }
      generalIssues {
        scwTitle
        scwDescription
        scwId
        issues {
          id
          confidence
          impact
          description
          snippet
          start
          end
          additionalData {
            title
            description
          }
          governanceInfo {
            owners {
              timelockDelay
              type
              owner
            }
            worstOwner {
              timelockDelay
              type
              owner
            }
          }
          severityChanges {
            from
            to
            reason
          }
        }
      }
      stats {
        percentage
        scammed
      }
      diffs {
        id
        address
        network
        name
        projectName
        score
        createdAt
      }
      proxyData {
        sourceCodeLink
        proxyOwner
        proxyIssues {
          scwTitle
          scwDescription
          scwId
          issues {
            id
            impact
            description
            snippet
            start
            end
            additionalData {
              title
              description
            }
            governanceInfo {
              owners {
                timelockDelay
                type
                owner
              }
              worstOwner {
                timelockDelay
                type
                owner
              }
            }
            severityChanges {
              from
              to
              reason
            }
          }
        }
        implementationData {
          firstTxFrom
          firstTxDate
          firstTxBlock
          name
          initialFunder
          initialFunding
        }
      }
      governance {
        visibleOwner
        proxyOwners {
          type
          owner
          timelockDelay
          timelock
          modifiable
          impact
          governance {
            proposals
            proposalMaxActions
            votingPeriod
            quorum
            threshold
            name
          }
          multisig {
            threshold
            transactionCount
            multisigOwners
          }
        }
        contractOwners {
          type
          owner
          timelockDelay
          timelock
          modifiable
          impact
          governance {
            proposals
            proposalMaxActions
            votingPeriod
            quorum
            threshold
            name
          }
          multisig {
            threshold
            transactionCount
            multisigOwners
          }
        }
        issueOwners {
          scwId
          owners {
            type
            owner
            timelockDelay
            timelock
            modifiable
            impact
            governance {
              proposals
              proposalMaxActions
              votingPeriod
              quorum
              threshold
              name
            }
            multisig {
              threshold
              transactionCount
              multisigOwners
            }
          }
        }
      }
    }
  }`;

const governanceInfo_query = `
  query($address: String!, $network: Int!) {
    governanceInfo(address: $address, network: $network) {
      snapshot {
        id name
        followers
        link
        proposalsCount
        proposalsCount7d
        votesCount
        votesCount7d
        membersCount
        strategiesCount
        proposalsActive
        proposalsFailed
        proposalsPassed
        quorum
        members {
          address
          type
        }
        strategies {
          name
          description
          additionalInfo
        }
        proposals {
          title
          state
          winningChoice
          created
          votes
          start
          end
          link
        }
        proposalsAllTime {
          successful
          date
          votesCount
        }
      }
      tally {
        id name
        followers
        link
        proposalsCount
        proposalsCount7d
        votesCount
        votesCount7d
        membersCount
        strategiesCount
        proposalsActive
        proposalsFailed
        proposalsPassed
        quorum
        members {
          address
          type
        }
        strategies {
          name
          description
          additionalInfo
        }
        proposals {
          title
          state
          winningChoice
          created
          votes
          start
          end
          link
        }
        proposalsAllTime {
          successful
          date
          votesCount
        }
      }
    }
  }`     
  
const fakeTokenDetection_query = `
  query($address: String!, $network: Int!) {
    fakeTokenDetection(address: $address, name: "", symbol: "", networkId: $network, simplified: true){
      fake
      originalTokenLink
    }
  }`

const liquidityAnalysis_query = `
  query($address: String!, $network: Int!) {
    liquidityAnalysis(address: $address, network: $network) {
      sell24h
      buy24h
      totalLiquidity
      totalLockedPercent
      totalBurnedPercent
      totalCreatorPercent
      totalUnlockedPercent
      lpHolders {
        address
        liquidityUsd
        isContract
        locker
        isVerified
        unlockDate
        percentage
        tokensAmount
        locker
      }
      liquidityPools {
        initialLiquidity {
          token
          amount
          amountUsd
        }
        liquidityDistribution {
          creator
          creatorPercentage
          locked
          lockedPercentage
          burned
          burnedPercentage
        }
        name
        source
        address
        tokens {
          price
          reserve
          reserveUsd
          symbol
          address
          logo
        }
        isAdequateLiquidityPresent
        isEnoughLiquidityLocked
        isCreatorNotContainLiquidity
        liquidityUsd
        totalWeigth
        swapFee
        owner
      }
      issues {
        scwId
        scwTitle
        scwDescription
        issues {
          impact
          id
          confidence
          description
          start
          end
          data
          snippet
          severityChanges {
            from
            reason
            to
          }
          governanceInfo {
            owners {
              type
              owner
              modifiable
              multisig {
                multisigOwners
                threshold
                transactionCount
              }
              governance {
                proposals
                proposalMaxActions
                votingPeriod
                quorum
                threshold
                name
              }
              timelockDelay
              timelock
            }
            worstOwner {
              type
              owner
              modifiable
              multisig {
                multisigOwners
                threshold
                transactionCount
              }
              governance {
                proposals
                proposalMaxActions
                votingPeriod
                quorum
                threshold
                name
              }
              timelockDelay
              timelock
            }
          }
        }
      }
    }
  }`

const holderAnalysis_query = `
  query($address: String!, $network: Int!) {
    holderAnalysis(address: $address, network: $network, includeHolderType: true) {
      topHolders {
        address
        balance
        percent
        isContract
        isVerified
      }
      lockedPercentage
      topHoldersTotal
      topHoldersTotalPercentage
      creator
      creatorBalance
      creatorBalancePercentage
      owner
      ownerBalance
      ownerBalancePercentage
      burned
      burnedPercentage
      tokenTotalSupply
      totalHolders
      issues {
        scwId
        scwTitle
        scwDescription
        issues {
          impact
          id
          confidence
          description
          start
          end
          data
          snippet
          severityChanges {
            from
            reason
            to
          }
          governanceInfo {
            owners {
              type
              owner
              modifiable
              multisig {
                multisigOwners
                threshold
                transactionCount
              }
              governance {
                proposals
                proposalMaxActions
                votingPeriod
                quorum
                threshold
                name
              }
              timelockDelay
              timelock
            }
            worstOwner {
              type
              owner
              modifiable
              multisig {
                multisigOwners
                threshold
                transactionCount
              }
              governance {
                proposals
                proposalMaxActions
                votingPeriod
                quorum
                threshold
                name
              }
              timelockDelay
              timelock
            }
          }
        }
      }
    }
  }`

const socialAnalysis_query = `
  query($address: String!, $network: Int!) {
    socialAnalysis(address: $address, network: $network) {
      isWebsiteFound
      isWebsiteAlive
      isWebsiteFound
      isAddressConnectedToWebsite
      websiteLinks
      score
      redditInfo {
        url
        subscribers
        createdAt
        positiveCount
        negativeCount
        neutralCount
      }
      telegramInfo {
        name
        messages
        createdAt
        subscribers
      }
      issues {
        scwId
        scwTitle
        scwDescription
        issues {
          id
          impact
          confidence
          description
          start
          end
          data
          snippet
          severityChanges {
            from
            reason
            to
          }
        }
      }
    }
  }`

const score_query = `
  query($address: String!, $network: Int!) {
    score(address: $address, network: $network) {
      score
      whitelisted
      exploited
      dimensionsAmount
      finalResult
      dimensions {
        score
        name
      }
    }
  }`;

const variables = {
    "address": address,
    "network": chain_id
  };

(async () => {
  let returnValue = {}
  if (project_request === "T"){
    const project_data = await fetchGraphQLData(project_query, variables);
    returnValue["project"] = project_data.data.project
  } else {
    returnValue["project"] = false
  }

  if (liquidity_request === "T"){
    const liquidityAnalysis_data = await fetchGraphQLData(liquidityAnalysis_query, variables);
    returnValue["liquidityAnalysis"] = liquidityAnalysis_data.data.liquidityAnalysis
  } else {
    returnValue["liquidityAnalysis"] = false
  }

  if (holder_request === "T"){
    const holderAnalysis_data = await fetchGraphQLData(holderAnalysis_query, variables);
    returnValue["holderAnalysis"] = holderAnalysis_data.data.holderAnalysis
  } else {
    returnValue["holderAnalysis"] = false
  }

  if (score_request === "T"){
    const score_data = await fetchGraphQLData(score_query, variables);
    returnValue["score"] = score_data.data.score
  } else {
    returnValue["score"] = false
  }

  console.log(JSON.stringify(returnValue)); // Convert the object to a JSON string and print it
  process.exit(0);
})();